worker/uniter/container/resolver.go

```diff
@@ -0,0 +1,71 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package container
+
+import (
+	"github.com/juju/loggo"
+
+	"github.com/juju/juju/worker/uniter/operation"
+	"github.com/juju/juju/worker/uniter/remotestate"
+	"github.com/juju/juju/worker/uniter/resolver"
+)
+
+var logger = loggo.GetLogger("juju.worker.uniter.actions")
+
+type containerResolver struct{}
+
+// NewResolver returns a new resolver with determines which action related operation
+// should be run based on local and remote uniter states.
+func NewResolver() resolver.Resolver {
+	return &containerResolver{}
+}
+
+// NextOp implements the resolver.Resolver interface.
+func (r *containerResolver) NextOp(
+	localState resolver.LocalState,
+	remoteState remotestate.Snapshot,
+	opFactory operation.Factory,
+) (operation.Operation, error) {
+	noOp := func() (operation.Operation, error) {
+		if localState.Kind == operation.RemoteInit {
+			// If we are resuming from an unexpected state, skip init.
+			// Retry will occur when remotestate updates.
+			return opFactory.NewSkipRemoteInit(false)
+		}
+		return nil, resolver.ErrNoOperation
+	}
+	if remoteState.RunningStatus == nil {
+		return noOp()
+	}
+	// Check if init or workload containers are running.
+	if !remoteState.RunningStatus.Initialising &&
+		!remoteState.RunningStatus.Running {
+		return noOp()
+	}
+	// If we haven't yet handled the init container.
+	if !localState.OutdatedRemoteCharm && localState.RunningStatus != nil {
+		if localState.RunningStatus.InitialisingTime == remoteState.RunningStatus.InitialisingTime {
+			// We've already initilised the container.
+			return noOp()
+		}
+	} else if !localState.OutdatedRemoteCharm {
+		// Nothing to do
+		return noOp()
+	}
+	switch localState.Kind {
+	case operation.RunHook:
+		if localState.Step == operation.Pending {
+			return opFactory.NewRemoteInit(*remoteState.RunningStatus)
```

Comment:
we only want to do this if remote init has not yet been done and/or requires doing.
will this achieve that? if seems if there's any pending hook at all we always run remote init?

Reply:
yep, we can check resolver localstate, but will have the same outcome because resuming from saved operation state happens if the resolver restarted and when we restart we always assume the remote state needs init

Reply:
I'm a little confused. It seems this will result in an endless loop? If we get a pending hook, we'll always do remote init, then the top level resolver loop starts again and we'll do remote init....
(remote init commit function writes out hook pending)

Reply:
After a remote init operation succeeds, the resolver updates the resolver local state.

```
	return onCommitWrapper{op, func(*operation.State) {
		s.LocalState.RunningStatus = &runningStatus
		s.LocalState.OutdatedRemoteCharm = false
	}}, nil
```

So both the running status has been consumed (so the init container has finished and the pod has the charm) and we have marked the charm as being up-to date.